File: Distribute_candies_in_a_binary_Tree.py
class Solution:
    def solve(self, root):
        if root is None:
            return 0, 0
        left_profit, a1 = self.solve(root.left)
        right_profit, a2 = self.solve(root.right)
        ans = abs(left_profit) + abs(right_profit)
        return root.data + left_profit + right_profit - 1, ans + a1 + a2
    def distributeCandy(self, root):
        _, ans = self.solve(root)
        return ans

File: Distribute_candies_in_a_binary_Tree.java
class Solution{
    static int ans;
    static int solve(Node root){
        if(root == null){
            return 0;
        }
        int left_profit = solve(root.left);
        int right_profit = solve(root.right);
        ans += Math.abs(left_profit) + Math.abs(right_profit);
        return root.data + left_profit + right_profit - 1;
    }
    public static int distributeCandy(Node root){
        ans = 0;
        solve(root);
        return ans;
    }
}
